File: backend/app/schemas/audience.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class Contact(BaseModel):
    id: str
    kind: str
    name: str
    company: Optional[str] = None
    email: Optional[str] = None
    country: Optional[str] = None
    location: Optional[str] = None
    tags: List[str] = []
    meta: str
    source: str
    created_at: Optional[str] = None


class ContactPage(BaseModel):
    items: List[Contact]
    total: int


class AudienceOverview(BaseModel):
    contacts: int
    leads: int
    customers: int
    followers_total: int
    by_platform: List[Dict[str, Any]]
    by_country: List[Any]
    by_source: Dict[str, int]


class SegmentCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=160)
    description: Optional[str] = None
    source: str = "all"
    filters: Dict[str, Any] = {}
    color: Optional[str] = None


class SegmentResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    source: str
    filters: Dict[str, Any]
    color: Optional[str] = None
    count: int = 0
    workspace_id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SegmentList(BaseModel):
    items: List[SegmentResponse]

File: frontend/lib/api/audience.ts
import { http, qs } from "./http";
import type { PlatformRow } from "./analytics";

export interface Contact {
  id: string;
  kind: "lead" | "customer";
  name: string;
  company: string | null;
  email: string | null;
  country: string | null;
  location: string | null;
  tags: string[];
  meta: string;
  source: string;
  created_at: string | null;
}

export interface AudienceOverview {
  contacts: number;
  leads: number;
  customers: number;
  followers_total: number;
  by_platform: PlatformRow[];
  by_country: [string, number][];
  by_source: Record<string, number>;
}

export interface Segment {
  id: string;
  name: string;
  description: string | null;
  source: string;
  filters: Record<string, string>;
  color: string | null;
  count: number;
  workspace_id: string;
  created_at: string;
  updated_at: string;
}

export interface ContactFilters {
  source?: string;
  country?: string;
  tag?: string;
  meta?: string;
  search?: string;
  limit?: number;
  offset?: number;
}

export const audienceApi = {
  overview: () => http.get<AudienceOverview>("/audience/overview"),
  contacts: (f: ContactFilters = {}) =>
    http.get<{ items: Contact[]; total: number }>(`/audience/contacts${qs({ ...f })}`),
  segments: () => http.get<{ items: Segment[] }>("/audience/segments"),
  createSegment: (body: { name: string; source: string; filters: Record<string, string> }) =>
    http.post<Segment>("/audience/segments", body),
  deleteSegment: (id: string) => http.del<void>(`/audience/segments/${id}`),
};
